course/page/code.py

```diff
@@ -292,6 +294,97 @@ def request_python_run_with_retries(run_req, run_timeout, image=None, retry_coun
         return result
 
 
+# {{{ HTML sanitization helpers
+
+def is_allowed_data_uri(allowed_mimetypes, uri):
+    import re
+    m = re.match(r"^data:([-a-z0-9]+/[-a-z0-9]+);base64,", uri)
+    if not m:
+        return False
+
+    mimetype = m.group(1)
+    return mimetype in allowed_mimetypes
+
+
+def filter_from_code_html_attributes(tag, name, value):
+    from html5lib.filters.sanitizer import attr_val_is_uri
+    if tag == "audio":
+        if name in ["controls"]:
+            return True
+        else:
+            return False
+
+    elif tag == "source":
+        if name in ["type"]:
+            return True
+        elif name == "src":
+            if is_allowed_data_uri([
+                    "audio/wav",
+                    ], value):
+                return True
+            else:
+                return False
+        else:
+            return False
+
+    elif tag == "img":
+        if name in ["alt", "title"]:
+            return True
+        elif name == "src":
+            if is_allowed_data_uri([
+                    "image/png",
+                    "image/jpeg",
+                    ], value):
+                return True
+        else:
+            return False
+
+    else:
+        # Approach recommended here: https://github.com/mozilla/bleach/issues/348
+
+        # FIXME: Cannot access 'namespaced name' of attribute, so
+        # we're unable to match XML's base and xlink.
+        if (None, name) in attr_val_is_uri:
+            # Reluctant copy-paste from
+            # https://github.com/mozilla/bleach/blob/8706b5373b633407e804b1b2975edf8760067ff7/bleach/sanitizer.py#L513-L525
+
+            from xml.sax.saxutils import unescape
+
+            val_unescaped = re.sub(
+                "[`\000-\040\177-\240\s]+",
+                '',
+                unescape(value)).lower()
+
+            # Remove replacement characters from unescaped characters
+            val_unescaped = val_unescaped.replace("\ufffd", "")
+
+            # Drop attributes with uri values that have
+            # protocols that aren't allowed
+            if (re.match(r'^[a-z0-9][-+.a-z0-9]*:', val_unescaped)
+                    and
+                    (val_unescaped.split(':')[0]
+                        not in bleach.ALLOWED_PROTOCOLS)):
+                return False
```

Comment:
My comment above say, the logic here is redudant. The reason is it has no difference with the bleach logic, just repeating. That's not what you've expected. So we need to make it different. 

Let me guess: We are allowing only 'img', 'video' and 'source' to have uri attribute? If that's true, it's quite simple then, just fail those which have uri attribute is ok.

Reply:
I simple strip `tags` which is not allowed, I think that will meet your requirements.